### ISSUE TITLE
[elasticsearch] Elasticsearch master only is failing with `UnavailableShardsException`

### ISSUE THREAD

**Issue by jmlrt** [elastic/helm-charts#1462]
**Chart version:** [7.x branch](https://github.com/elastic/helm-charts/tree/7.x) with `7.16.0-SNAPSHOT` images

**Kubernetes version:** reproduced on 1.19-1.21

**Kubernetes provider:** E.g. GKE (Google Kubernetes Engine) reproduced on GKE

**Helm Version:** reproduced with 3.6.2

**`helm get release` output**

e.g. `helm get elasticsearch` (replace `elasticsearch` with the name of your helm release)

*Be careful to obfuscate every secrets (credentials, token, public IP, ...) that could be visible in the output before copy-pasting.*

*If you find some secrets in plain text in `helm get release` output you should use [Kubernetes Secrets](https://kubernetes.io/docs/concepts/configuration/secret/) to managed them is a secure way (see [Security Example](https://github.com/elastic/helm-charts/blob/master/elasticsearch/examples/security/values.yaml#L23-L38)).*

<details>
<summary>Output of helm get release</summary>

```
NAME: helm-es-multi-master
LAST DEPLOYED: Mon Dec 13 10:26:24 2021
NAMESPACE: jmlrt-test
STATUS: pending-install
REVISION: 1
USER-SUPPLIED VALUES:
clusterName: multi
nodeGroup: master
roles:
  data: "false"
  ingest: "false"
  master: "true"
  ml: "false"
  remote_cluster_client: "false"

COMPUTED VALUES:
antiAffinity: hard
antiAffinityTopologyKey: kubernetes.io/hostname
clusterHealthCheckParams: wait_for_status=green&timeout=1s
clusterName: multi
enableServiceLinks: true
envFrom: []
esConfig: {}
esJavaOpts: ""
esMajorVersion: ""
extraContainers: []
extraEnvs: []
extraInitContainers: []
extraVolumeMounts: []
extraVolumes: []
fsGroup: ""
fullnameOverride: ""
healthNameOverride: ""
hostAliases: []
httpPort: 9200
image: docker.elastic.co/elasticsearch/elasticsearch
imagePullPolicy: IfNotPresent
imagePullSecrets: []
imageTag: 7.16.0-SNAPSHOT
ingress:
  annotations: {}
  enabled: false
  hosts:
  - host: chart-example.local
    paths:
    - path: /
  tls: []
initResources: {}
keystore: []
labels: {}
lifecycle: {}
masterService: ""
maxUnavailable: 1
minimumMasterNodes: 2
nameOverride: ""
networkHost: 0.0.0.0
networkPolicy:
  http:
    enabled: false
  transport:
    enabled: false
nodeAffinity: {}
nodeGroup: master
nodeSelector: {}
persistence:
  annotations: {}
  enabled: true
  labels:
    enabled: false
podAnnotations: {}
podManagementPolicy: Parallel
podSecurityContext:
  fsGroup: 1000
  runAsUser: 1000
podSecurityPolicy:
  create: false
  name: ""
  spec:
    fsGroup:
      rule: RunAsAny
    privileged: true
    runAsUser:
      rule: RunAsAny
    seLinux:
      rule: RunAsAny
    supplementalGroups:
      rule: RunAsAny
    volumes:
    - secret
    - configMap
    - persistentVolumeClaim
    - emptyDir
priorityClassName: ""
protocol: http
rbac:
  automountToken: true
  create: false
  serviceAccountAnnotations: {}
  serviceAccountName: ""
readinessProbe:
  failureThreshold: 3
  initialDelaySeconds: 10
  periodSeconds: 10
  successThreshold: 3
  timeoutSeconds: 5
replicas: 3
resources:
  limits:
    cpu: 1000m
    memory: 2Gi
  requests:
    cpu: 1000m
    memory: 2Gi
roles:
  data: "false"
  ingest: "false"
  master: "true"
  ml: "false"
  remote_cluster_client: "false"
schedulerName: ""
secretMounts: []
securityContext:
  capabilities:
    drop:
    - ALL
  runAsNonRoot: true
  runAsUser: 1000
service:
  annotations: {}
  enabled: true
  externalTrafficPolicy: ""
  httpPortName: http
  labels: {}
  labelsHeadless: {}
  loadBalancerIP: ""
  loadBalancerSourceRanges: []
  nodePort: ""
  transportPortName: transport
  type: ClusterIP
sysctlInitContainer:
  enabled: true
sysctlVmMaxMapCount: 262144
terminationGracePeriod: 120
tests:
  enabled: true
tolerations: []
transportPort: 9300
updateStrategy: RollingUpdate
volumeClaimTemplate:
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: 30Gi

HOOKS:
---
# Source: elasticsearch/templates/test/test-elasticsearch-health.yaml
apiVersion: v1
kind: Pod
metadata:
  name: "helm-es-multi-master-ocmaq-test"
  annotations:
    "helm.sh/hook": test
    "helm.sh/hook-delete-policy": hook-succeeded
spec:
  securityContext:
    fsGroup: 1000
    runAsUser: 1000
  containers:
  - name: "helm-es-multi-master-vcsfq-test"
    image: "docker.elastic.co/elasticsearch/elasticsearch:7.16.0-SNAPSHOT"
    imagePullPolicy: "IfNotPresent"
    command:
      - "sh"
      - "-c"
      - |
        #!/usr/bin/env bash -e
        curl -XGET --fail 'multi-master:9200/_cluster/health?wait_for_status=green&timeout=1s'
  restartPolicy: Never
MANIFEST:
---
# Source: elasticsearch/templates/poddisruptionbudget.yaml
apiVersion: policy/v1beta1
kind: PodDisruptionBudget
metadata:
  name: "multi-master-pdb"
spec:
  maxUnavailable: 1
  selector:
    matchLabels:
      app: "multi-master"
---
# Source: elasticsearch/templates/service.yaml
kind: Service
apiVersion: v1
metadata:
  name: multi-master
  labels:
    heritage: "Helm"
    release: "helm-es-multi-master"
    chart: "elasticsearch"
    app: "multi-master"
  annotations:
    {}
spec:
  type: ClusterIP
  selector:
    release: "helm-es-multi-master"
    chart: "elasticsearch"
    app: "multi-master"
  ports:
  - name: http
    protocol: TCP
    port: 9200
  - name: transport
    protocol: TCP
    port: 9300
---
# Source: elasticsearch/templates/service.yaml
kind: Service
apiVersion: v1
metadata:
  name: multi-master-headless
  labels:
    heritage: "Helm"
    release: "helm-es-multi-master"
    chart: "elasticsearch"
    app: "multi-master"
  annotations:
    service.alpha.kubernetes.io/tolerate-unready-endpoints: "true"
spec:
  clusterIP: None # This is needed for statefulset hostnames like elasticsearch-0 to resolve
  # Create endpoints also if the related pod isn't ready
  publishNotReadyAddresses: true
  selector:
    app: "multi-master"
  ports:
  - name: http
    port: 9200
  - name: transport
    port: 9300
---
# Source: elasticsearch/templates/statefulset.yaml
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: multi-master
  labels:
    heritage: "Helm"
    release: "helm-es-multi-master"
    chart: "elasticsearch"
    app: "multi-master"
  annotations:
    esMajorVersion: "7"
spec:
  serviceName: multi-master-headless
  selector:
    matchLabels:
      app: "multi-master"
  replicas: 3
  podManagementPolicy: Parallel
  updateStrategy:
    type: RollingUpdate
  volumeClaimTemplates:
  - metadata:
      name: multi-master
    spec:
      accessModes:
      - ReadWriteOnce
      resources:
        requests:
          storage: 30Gi
  template:
    metadata:
      name: "multi-master"
      labels:
        release: "helm-es-multi-master"
        chart: "elasticsearch"
        app: "multi-master"
      annotations:
        
    spec:
      securityContext:
        fsGroup: 1000
        runAsUser: 1000
      automountServiceAccountToken: true
      affinity:
        podAntiAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
          - labelSelector:
              matchExpressions:
              - key: app
                operator: In
                values:
                - "multi-master"
            topologyKey: kubernetes.io/hostname
      terminationGracePeriodSeconds: 120
      volumes:
      enableServiceLinks: true
      initContainers:
      - name: configure-sysctl
        securityContext:
          runAsUser: 0
          privileged: true
        image: "docker.elastic.co/elasticsearch/elasticsearch:7.16.0-SNAPSHOT"
        imagePullPolicy: "IfNotPresent"
        command: ["sysctl", "-w", "vm.max_map_count=262144"]
        resources:
          {}

      containers:
      - name: "elasticsearch"
        securityContext:
          capabilities:
            drop:
            - ALL
          runAsNonRoot: true
          runAsUser: 1000
        image: "docker.elastic.co/elasticsearch/elasticsearch:7.16.0-SNAPSHOT"
        imagePullPolicy: "IfNotPresent"
        readinessProbe:
          exec:
            command:
              - bash
              - -c
              - |
                #!/usr/bin/env bash -e
                # If the node is starting up wait for the cluster to be ready (request params: "wait_for_status=green&timeout=1s" )
                # Once it has started only check that the node itself is responding
                START_FILE=/tmp/.es_start_file

                # Disable nss cache to avoid filling dentry cache when calling curl
                # This is required with Elasticsearch Docker using nss < 3.52
                export NSS_SDB_USE_CACHE=no

                http () {
                  local path="${1}"
                  local args="${2}"
                  set -- -XGET -s

                  if [ "$args" != "" ]; then
                    set -- "$@" $args
                  fi

                  if [ -n "${ELASTIC_USERNAME}" ] && [ -n "${ELASTIC_PASSWORD}" ]; then
                    set -- "$@" -u "${ELASTIC_USERNAME}:${ELASTIC_PASSWORD}"
                  fi

                  curl --output /dev/null -k "$@" "http://127.0.0.1:9200${path}"
                }

                if [ -f "${START_FILE}" ]; then
                  echo 'Elasticsearch is already running, lets check the node is healthy'
                  HTTP_CODE=$(http "/" "-w %{http_code}")
                  RC=$?
                  if [[ ${RC} -ne 0 ]]; then
                    echo "curl --output /dev/null -k -XGET -s -w '%{http_code}' \${BASIC_AUTH} http://127.0.0.1:9200/ failed with RC ${RC}"
                    exit ${RC}
                  fi
                  # ready if HTTP code 200, 503 is tolerable if ES version is 6.x
                  if [[ ${HTTP_CODE} == "200" ]]; then
                    exit 0
                  elif [[ ${HTTP_CODE} == "503" && "7" == "6" ]]; then
                    exit 0
                  else
                    echo "curl --output /dev/null -k -XGET -s -w '%{http_code}' \${BASIC_AUTH} http://127.0.0.1:9200/ failed with HTTP code ${HTTP_CODE}"
                    exit 1
                  fi

                else
                  echo 'Waiting for elasticsearch cluster to become ready (request params: "wait_for_status=green&timeout=1s" )'
                  if http "/_cluster/health?wait_for_status=green&timeout=1s" "--fail" ; then
                    touch ${START_FILE}
                    exit 0
                  else
                    echo 'Cluster is not yet ready (request params: "wait_for_status=green&timeout=1s" )'
                    exit 1
                  fi
                fi
          failureThreshold: 3
          initialDelaySeconds: 10
          periodSeconds: 10
          successThreshold: 3
          timeoutSeconds: 5
        ports:
        - name: http
          containerPort: 9200
        - name: transport
          containerPort: 9300
        resources:
          limits:
            cpu: 1000m
            memory: 2Gi
          requests:
            cpu: 1000m
            memory: 2Gi
        env:
          - name: node.name
            valueFrom:
              fieldRef:
                fieldPath: metadata.name
          - name: cluster.initial_master_nodes
            value: "multi-master-0,multi-master-1,multi-master-2,"
          - name: discovery.seed_hosts
            value: "multi-master-headless"
          - name: cluster.name
            value: "multi"
          - name: network.host
            value: "0.0.0.0"
          - name: node.data
            value: "false"
          - name: node.ingest
            value: "false"
          - name: node.master
            value: "true"
          - name: node.ml
            value: "false"
          - name: node.remote_cluster_client
            value: "false"
        volumeMounts:
          - name: "multi-master"
            mountPath: /usr/share/elasticsearch/data

NOTES:
1. Watch all cluster members come up.
  $ kubectl get pods --namespace=jmlrt-test -l app=multi-master -w2. Test cluster health using Helm test.
  $ helm --namespace=jmlrt-test test helm-es-multi-master
```

</details>

**Describe the bug:**

Deploying Elasticsearch master only is failing with `UnavailableShardsException`

**Steps to reproduce:**

1. checkout 7.x branch
2. Deploy Elasticsearch chart with the following values:
```
clusterName: "multi"
nodeGroup: "master"
imageTag: "7.16.0-SNAPSHOT"
roles:
  master: "true"
  ingest: "false"
  data: "false"
  ml: "false"
  remote_cluster_client: "false"
```

**Expected behavior:**

The Helm chart should deploy successfully and we should have 3 master node with RUNNING state.

**Provide logs and/or server output (if relevant):**

```
elasticsearch {"type": "server", "timestamp": "2021-12-09T17:44:44,872Z", "level": "ERROR", "component": "o.e.x.i.h.ILMHistoryStore", "cluster.name": "multi", "node.name": "multi-master-0", "message": "failures: [{awVLoH0B_9bA
4dNPOwvn=UnavailableShardsException[[.ds-ilm-history-5-2021.12.09-000001][0] primary shard is not active Timeout: [1m], request: [BulkShardRequest [[.ds-ilm-history-5-2021.12.09-000001][0]] containing [5] requests]], agVLoH0B_
9bA4dNPOwvn=UnavailableShardsException[[.ds-ilm-history-5-2021.12.09-000001][0] primary shard is not active Timeout: [1m], request: [BulkShardRequest [[.ds-ilm-history-5-2021.12.09-000001][0]] containing [5] requests]], bAVLoH
0B_9bA4dNPOwvn=UnavailableShardsException[[.ds-ilm-history-5-2021.12.09-000001][0] primary shard is not active Timeout: [1m], request: [BulkShardRequest [[.ds-ilm-history-5-2021.12.09-000001][0]] containing [5] requests]], aQV
LoH0B_9bA4dNPOwvn=UnavailableShardsException[[.ds-ilm-history-5-2021.12.09-000001][0] primary shard is not active Timeout: [1m], request: [BulkShardRequest [[.ds-ilm-history-5-2021.12.09-000001][0]] containing [5] requests]],
aAVLoH0B_9bA4dNPOwvn=UnavailableShardsException[[.ds-ilm-history-5-2021.12.09-000001][0] primary shard is not active Timeout: [1m], request: [BulkShardRequest [[.ds-ilm-history-5-2021.12.09-000001][0]] containing [5] requests]
]}]", "cluster.uuid": "VBvbOCHJQXSpiq_2s9kj4A", "node.id": "oBfMyeTPR0ChJQMKZukcRQ"  }
```

**Any additional context:**
Same test with replacing `imageTag: 7.16.0-SNAPSHOT` by `imageTag: 7.15.2` is successful.

cc @elastic/es-delivery @elastic/infra-release

**Comment by jmlrt** [comment]
Attaching the full logs of the 3 Elasticsearch nodes:

[es-master-0.log](https://github.com/elastic/helm-charts/files/7704833/es-master-0.log)
[es-master-1.log](https://github.com/elastic/helm-charts/files/7704834/es-master-1.log)
[es-master-2.log](https://github.com/elastic/helm-charts/files/7704835/es-master-2.log)

**Comment by pugnascotia** [comment]
I've asked @pgomulka to have a look as on first glance it maybe related to https://github.com/elastic/elasticsearch/pull/80406.

**Comment by pugnascotia** [comment]
I had missed the fact that this test case consist purely of master-only nodes, meaning that when Elasticsearch tries to create a data stream for writing deprecation logs, it can't and so it fails. We should be able to disable `cluster.deprecation_indexing.enabled` in order to stop the data stream being created, which should let the test pass at least.

**Comment by jmlrt** [comment]
Closed in #1474
